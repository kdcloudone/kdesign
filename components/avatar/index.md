---
category: Components
type: 数据展示
subtitle: 头像
title: Avatar
order: 0
---

头像是一种代表人或物的组件。

## 使用场景

需要代表某个人或物时使用，例如：用户、商家、店铺和企业等。

## API

| 属性 | 说明 | 类型 | 默认值 | 版本 |
| --- | --- | --- | --- | --- |
| alt | 图像无法显示时的替代文本 | string | - | 1.0.0 |
| draggable | 图片是否允许拖动 | boolean \| `'true'` \| `'false'` | - | 1.0.0 |
| gap | 字符类型距离左右两侧边界单位像素 | number | 4 | 1.0.0 |
| icon | 设置头像的自定义图标 | ReactNode | - | 1.0.0 |
| onError | 图片加载失败的事件，返回 false 会关闭组件默认的 fallback 行为 | () => boolean | - | 1.0.0 |
| shape | 指定头像的形状 | `circle` \| `square` | `circle` | 1.0.0 |
| size | 设置头像的大小 | number \| `large` \| `middle` \| `small` \| `extra-small` | `middle` | 1.0.0 |
| src | 图片类头像的资源地址或者图片元素 | string \| ReactNode | - | 1.0.0 |
| srcSet | 设置图片类头像响应式资源地址 | string | - | 1.0.0 |
