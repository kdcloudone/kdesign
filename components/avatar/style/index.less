@import '../../style/themes/index';
@import './mixin.less';
@avatar-prefix-cls: ~'@{kd-prefix}-avatar';
@icon-prefix-cls: ~'@{kd-prefix}icon';

.@{avatar-prefix-cls} {
  position: relative;
  display: inline-block;
  overflow: hidden;
  white-space: nowrap;
  text-align: center;
  vertical-align: middle;
  background: @avatar-bg;

  &-image {
    background: transparent;
  }

  &-disabled {
    cursor: not-allowed;
  }

  &:hover:not(&-disabled) {
    cursor: pointer;
  }

  .avatar-size(@avatar-size-base, @avatar-font-size-base);

  &-lg {
    .avatar-size(@avatar-size-lg, @avatar-font-size-lg);
  }

  &-sm {
    .avatar-size(@avatar-size-sm, @avatar-font-size-sm);
  }

  &-xs {
    .avatar-size(@avatar-size-xs, @avatar-font-size-xs);
  }

  &-square {
    border-radius: @avatar-border-radius;
  }

  & > img {
    display: block;
    width: 100%;
    height: 100%;
    object-fit: cover;
  }
}

.avatar-size(@size, @font-size) {
  width: @size;
  height: @size;
  line-height: @size;
  border-radius: 50%;

  &-string {
    position: absolute;
    left: 50%;
    transform-origin: 0 center;
  }

  &.@{avatar-prefix-cls}-icon {
    font-size: @font-size;

    & > .@{icon-prefix-cls} {
      vertical-align: unset !important;
    }
  }
}