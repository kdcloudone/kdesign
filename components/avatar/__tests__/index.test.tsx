import React from 'react'
import { mount, render } from 'enzyme'
import ConfigProvider from '../../config-provider/index'
import Avatar from '../index'
import { AvatarShapes, AvatarShape, AvatarSizes } from '../avatar'
import Icon from '../../icon'
import mountTest from '../../../tests/shared/mountTest'

describe('Avatar', () => {
  // 1.mount test
  mountTest(Avatar)
  AvatarShapes.forEach((shape) => {
    mountTest(() => <Avatar shape={shape} />)
  })

  AvatarSizes.forEach((size) => {
    mountTest(() => <Avatar size={size} />)
  })

  // 2.render test
  it('renders correctly', () => {
    expect(render(<Avatar>Avatar</Avatar>)).toMatchSnapshot()

    AvatarSizes.forEach((size) => {
      expect(render(<Avatar size={size}>Avatar</Avatar>)).toMatchSnapshot()
    })
    AvatarShapes.forEach((shape) => {
      expect(render(<Avatar shape={shape}>Avatar</Avatar>)).toMatchSnapshot()
    })
  })

  // 3.warns in component
  it('warns if type is wrong', () => {
    const mockWarn = jest.fn()
    jest.spyOn(console, 'warn').mockImplementation(mockWarn)
    const shape = 'oval' as any as AvatarShape
    render(<Avatar shape={shape} />)
    expect(mockWarn).toHaveBeenCalledTimes(1)
    expect(mockWarn.mock.calls[0][0]).toMatch("Warning: [kdesign]-avatar: cannot found avatar shape 'oval'")
  })

  // 4. render null or undefined without errors
  it('render null or undefined buttons without errors', () => {
    expect(
      mount(
        <Avatar>
          {null}
          {undefined}
        </Avatar>,
      ),
    ).toMatchSnapshot()
  })

  // 5. displayName
  it('should have displayName static property', () => {
    const wrapper = mount(<Avatar>Avatar</Avatar>)
    expect((wrapper.type() as any).displayName).toBe('Avatar')
  })

  // 6. class state
  it('should class use right', () => {
    // default
    const DefaultAvatar = mount(<Avatar />)
    expect(DefaultAvatar.find('.kd-avatar')).toHaveClassName('.kd-avatar-md')
    expect(DefaultAvatar.find('.kd-avatar')).toHaveClassName('.kd-avatar-circle')

    const stringAvatar = mount(<Avatar>string</Avatar>)
    expect(stringAvatar.find('.kd-avatar').length).toBeGreaterThanOrEqual(1)

    AvatarShapes.forEach((shape) => {
      const TestButton = mount(<Avatar shape={shape}>Avatar</Avatar>)
      expect(TestButton.find(`.kd-avatar`)).toHaveClassName(`.kd-avatar-${shape}`)
    })

    const DisableAvatar = mount(<Avatar disabled />)
    expect(DisableAvatar.find(`.kd-avatar`)).toHaveClassName('.kd-avatar-disabled')

    const IconAvatar = mount(<Avatar icon={<Icon type="add"></Icon>} />)
    expect(IconAvatar.find(`.kd-avatar`)).toContainReact(<Icon type="add"></Icon>)
  })

  // 7.component interaction(event)
  it('Render long string correctly', () => {
    const wrapper = mount(<Avatar>TestString</Avatar>)
    const children = wrapper.find('.kd-avatar-string')
    expect(children.length).toBe(1)
  })

  it('should render fallback string correctly', () => {
    const div = global.document.createElement('div')
    global.document.body.appendChild(div)

    const wrapper = mount(<Avatar src="http://error.url">Fallback</Avatar>, { attachTo: div })
    wrapper.find('img').simulate('error')
    const children = wrapper.find('.kd-avatar-string')
    expect(children.length).toBe(1)
    expect(children.text()).toBe('Fallback')
    wrapper.detach()
    global.document.body.removeChild(div)
  })

  it('should handle onError correctly', () => {
    const LOAD_FAILURE_SRC = 'http://error.url'
    const LOAD_SUCCESS_SRC = 'https://zos.alipayobjects.com/rmsportal/ODTLcjxAfvqbxHnVXCYX.png'

    const div = global.document.createElement('div')
    global.document.body.appendChild(div)

    class Foo extends React.Component {
      state = {
        src: LOAD_FAILURE_SRC,
      }

      handleImgError = () => {
        this.setState({
          src: LOAD_SUCCESS_SRC,
        })
        return false
      }

      render() {
        const { src } = this.state
        return <Avatar src={src} onError={this.handleImgError} />
      }
    }

    const wrapper = mount(<Foo />, { attachTo: div })
    expect(div.querySelector('img')?.getAttribute('src')).toBe(LOAD_FAILURE_SRC)
    wrapper.find('img').simulate('error')

    expect(wrapper).toMatchSnapshot()
    expect(div.querySelector('img')?.getAttribute('src')).toBe(LOAD_SUCCESS_SRC)

    wrapper.detach()
    global.document.body.removeChild(div)
  })

  // 8.config provider
  it('should config use config provider', () => {
    const buttonConfig = {
      compDefaultProps: {
        Avatar: {
          size: 'large',
        },
      },
    }
    const wrapper = mount(
      <ConfigProvider value={buttonConfig}>
        <Avatar />
      </ConfigProvider>,
    )
    expect(wrapper.find('.kd-avatar')).toHaveClassName('.kd-avatar-lg')
  })

  // 9. ref test
  it('should get avatar element from ref', () => {
    const ref = React.createRef()
    mount(<Avatar ref={ref}></Avatar>)
    expect(ref.current instanceof HTMLSpanElement).toBe(true)
  })
})
