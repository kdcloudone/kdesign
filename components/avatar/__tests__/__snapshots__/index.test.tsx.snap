// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar render null or undefined buttons without errors 1`] = `
<Avatar>
  <span
    className="kd-avatar kd-avatar-md kd-avatar-circle"
    style={Object {}}
  >
    <span
      className="kd-avatar-string"
      style={
        Object {
          "transform": "scale(1) translateX(-50%)",
        }
      }
    />
  </span>
</Avatar>
`;

exports[`Avatar renders correctly 1`] = `
<span
  class="kd-avatar kd-avatar-md kd-avatar-circle"
>
  <span
    class="kd-avatar-string"
    style="opacity:0"
  >
    Avatar
  </span>
</span>
`;

exports[`Avatar renders correctly 2`] = `
<span
  class="kd-avatar kd-avatar-lg kd-avatar-circle"
>
  <span
    class="kd-avatar-string"
    style="opacity:0"
  >
    Avatar
  </span>
</span>
`;

exports[`Avatar renders correctly 3`] = `
<span
  class="kd-avatar kd-avatar-md kd-avatar-circle"
>
  <span
    class="kd-avatar-string"
    style="opacity:0"
  >
    Avatar
  </span>
</span>
`;

exports[`Avatar renders correctly 4`] = `
<span
  class="kd-avatar kd-avatar-sm kd-avatar-circle"
>
  <span
    class="kd-avatar-string"
    style="opacity:0"
  >
    Avatar
  </span>
</span>
`;

exports[`Avatar renders correctly 5`] = `
<span
  class="kd-avatar kd-avatar-xs kd-avatar-circle"
>
  <span
    class="kd-avatar-string"
    style="opacity:0"
  >
    Avatar
  </span>
</span>
`;

exports[`Avatar renders correctly 6`] = `
<span
  class="kd-avatar kd-avatar-md kd-avatar-circle"
>
  <span
    class="kd-avatar-string"
    style="opacity:0"
  >
    Avatar
  </span>
</span>
`;

exports[`Avatar renders correctly 7`] = `
<span
  class="kd-avatar kd-avatar-md kd-avatar-square"
>
  <span
    class="kd-avatar-string"
    style="opacity:0"
  >
    Avatar
  </span>
</span>
`;

exports[`Avatar should handle onError correctly 1`] = `
<Foo>
  <Avatar
    onError={[Function]}
    src="https://zos.alipayobjects.com/rmsportal/ODTLcjxAfvqbxHnVXCYX.png"
  >
    <span
      className="kd-avatar kd-avatar-md kd-avatar-circle kd-avatar-image"
      style={Object {}}
    >
      <img
        onError={[Function]}
        src="https://zos.alipayobjects.com/rmsportal/ODTLcjxAfvqbxHnVXCYX.png"
      />
    </span>
  </Avatar>
</Foo>
`;
